### PR TITLE
os/bluestore: avoid blob_t reencode when unchanged

### PR DESCRIPTION
**Pull request by liewegas**
This improves 4k random io iops by 9% on my nvme vstart cluster.

### CODE REVIEW

File: src/os/bluestore/BlueStore.h [hunk 1]
```diff
@@ -288,9 +288,14 @@ class BlueStore : public ObjectStore,
   struct Blob : public boost::intrusive::set_base_hook<> {
     std::atomic_int nref;  ///< reference count
     int64_t id = 0;          ///< id
+  private:
     bluestore_blob_t blob;   ///< blob metadata
+  public:
     BufferSpace bc;          ///< buffer cache
 
+    mutable bool dirty = true;
```
Comment (ifed01) [inline src/os/bluestore/BlueStore.h]: Can be private.

File: src/os/bluestore/BlueStore.h [hunk 2]
```diff
@@ -324,6 +340,22 @@ class BlueStore : public ObjectStore,
...
+    void encode(bufferlist& bl) const {
+      if (dirty) {
+	::encode(blob, blob_bl);
+	dirty = false;
+      } else {
+	assert(blob_bl.length());
+      }
+      bl.append(blob_bl);
+    }
+    void decode(bufferlist::iterator& p) {
```
Comment (ifed01) [inline src/os/bluestore/BlueStore.h]: I'm wondering if we can benefit from a deferred decoding. Just store encoded blob and increment a bufferlist iterator here. And do actual decode when get_blob or dirty_blob are called for the first time. The main question here is if there are decoded but unused blobs in the current implementation.

File: src/os/bluestore/BlueStore.cc [hunk 3]
```diff
@@ -3760,7 +3761,8 @@ int BlueStore::_do_read(
     } else {
       for (auto reg : b2r_it->second) {
 	// determine how much of the blob to read
-	uint64_t chunk_size = bptr->blob.get_chunk_size(csum_type != bluestore_blob_t::CSUM_NONE, block_size);
+	uint64_t chunk_size = bptr->dirty_blob().get_chunk_size(
```
Comment (ifed01) [inline src/os/bluestore/BlueStore.cc]: can be probably replaced with get_blob if get_chunk_size marked as const

File: src/os/bluestore/BlueStore.cc [hunk 4]
```diff
@@ -5931,7 +5936,10 @@ void BlueStore::_do_write_small(
 	     << " bstart 0x" << std::hex << bstart << std::dec << dendl;
 
     // can we pad our head/tail out with zeros?
-    uint64_t chunk_size = b->blob.get_chunk_size(b->blob.has_csum(), block_size); //blob csum settings to be applied hence ignoring current config settings for csum enable/disable
+    // blob csum settings to be applied hence ignoring current config
+    // settings for csum enable/disable
+    uint64_t chunk_size = b->dirty_blob().get_chunk_size(
```
Comment (ifed01) [inline src/os/bluestore/BlueStore.cc]: dirst_blob -> get_blob

File: src/os/bluestore/BlueStore.cc [hunk 5]
```diff
@@ -5931,7 +5936,10 @@ void BlueStore::_do_write_small(
...
+      b->get_blob().has_csum(), block_size);
```
Comment (ifed01) [inline src/os/bluestore/BlueStore.cc]: b->get_blob().has_csum() can be replaced with true here as has_csum is called internally

File: src/os/bluestore/BlueStore.h [hunk 6]
```diff
@@ -288,9 +288,14 @@ class BlueStore : public ObjectStore,
   struct Blob : public boost::intrusive::set_base_hook<> {
     std::atomic_int nref;  ///< reference count
     int64_t id = 0;          ///< id
+  private:
     bluestore_blob_t blob;   ///< blob metadata
+  public:
     BufferSpace bc;          ///< buffer cache
 
+    mutable bool dirty = true;
+    mutable bufferlist blob_bl;    ///< cached encoded blob
```
Comment (ifed01) [inline src/os/bluestore/BlueStore.h]: and this one too

**Comment by ifed01** [comment]
lgtm except mentioned minor issues

File: src/os/bluestore/BlueStore.h (hunk 2 above)
Comment (liewegas) [inline src/os/bluestore/BlueStore.h]: Good idea!  I suspect lots of blobs are unused and don't need to be decoded.

**Comment by ifed01** [comment]
@liewegas any performance difference for deferred decode?

**Comment by liewegas** [comment]
I didn't try to measure it.  Care to take a look?

File: src/os/bluestore/BlueStore.h [hunk 7]
```diff
@@ -314,6 +320,36 @@ class BlueStore : public ObjectStore,
       return out << b.id << ":" << b.blob;
```
Comment (ifed01) [inline src/os/bluestore/BlueStore.h]: b.get_blob() here?

**Comment by ifed01** [comment]
Just some knitpicking - the last commit contains some fixes unrelated to on-demand blob decoding. Moreover switch to the previous commit 45b757e results in build failures. That's a bit inconvenient if one wants to verify performance gain for the on-demand decode.

**Pull request merged**